cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/io/cdap/cdap/etl/spark/batch/RDDCollection.java

```diff
@@ -131,11 +182,20 @@ public RDDCollection(JavaSparkExecutionContext sec, JavaSparkContext jsc, SQLCon
       // which allows us to use a different number of partitions per joiner instead of using the global
       // spark.sql.shuffle.partitions setting in the spark conf
       if (joinPartitions != null && !toJoin.isBroadcast()) {
-        right = partitionOnKey(right, toJoin.getKey(), joinRequest.isNullSafe(), sparkSchema, joinPartitions);
+        List<String> rightKeys = new ArrayList<>(toJoin.getKey());
+        List<String> leftKeys = new ArrayList<>(joinRequest.getLeftKey());
+
+        // If distribution is enabled we need to add it to the partition keys to ensure we end up with the desired
+        // number of partitions
+        if (joinRequest.isDistributionEnabled() && JoinDistribution.getSupportedJoinTypes().contains(joinType)) {
```

Comment:
JoinDistribution should validate the join type earlier, which will let you remove the type check.